pkg/clients/subnetwork/subnetwork.go

```diff
@@ -42,28 +50,73 @@ func GenerateSubnetwork(s v1alpha3.SubnetworkParameters) *compute.Subnetwork {
-	s.Region = in.Region
-	for _, val := range in.SecondaryIpRanges {
-		obj := &v1alpha3.GCPSubnetworkSecondaryRange{
-			IPCidrRange: val.IpCidrRange,
+// GenerateSubnetworkForUpdate creates a *googlecompute.Subnetwork object using
+// SubnetworkParameters with fields disallowed by the GCP API removed. If a
+// field can be included in the GCP API but will result in an error if the value
+// is changed, it will still be included here such that users are notified of
+// invalid updates.
+func GenerateSubnetworkForUpdate(s v1beta1.Subnetwork, name string) *compute.Subnetwork {
+	sn := &compute.Subnetwork{
+		Name:                  name,
+		Description:           gcp.StringValue(s.Spec.ForProvider.Description),
+		EnableFlowLogs:        gcp.BoolValue(s.Spec.ForProvider.EnableFlowLogs),
+		IpCidrRange:           s.Spec.ForProvider.IPCidrRange,
+		PrivateIpGoogleAccess: gcp.BoolValue(s.Spec.ForProvider.PrivateIPGoogleAccess),
+		Fingerprint:           s.Status.AtProvider.Fingerprint,
+	}
+	for _, val := range s.Spec.ForProvider.SecondaryIPRanges {
+		obj := &compute.SubnetworkSecondaryRange{
+			IpCidrRange: val.IPCidrRange,
 			RangeName:   val.RangeName,
 		}
-		s.SecondaryIPRanges = append(s.SecondaryIPRanges, obj)
+		sn.SecondaryIpRanges = append(sn.SecondaryIpRanges, obj)
+	}
+	return sn
+}
+
+// GenerateSubnetworkObservation creates a SubnetworkObservation object using *googlecompute.Subnetwork.
+func GenerateSubnetworkObservation(in compute.Subnetwork) v1beta1.SubnetworkObservation {
+	return v1beta1.SubnetworkObservation{
+		CreationTimestamp: in.CreationTimestamp,
+		Fingerprint:       in.Fingerprint,
+		GatewayAddress:    in.GatewayAddress,
+		ID:                in.Id,
+		SelfLink:          in.SelfLink,
+	}
+}
+
+// LateInitializeSpec fills unassigned fields with the values in compute.Subnetwork object.
+func LateInitializeSpec(spec *v1beta1.SubnetworkParameters, in compute.Subnetwork) {
+	if spec.IPCidrRange == "" {
+		spec.IPCidrRange = in.IpCidrRange
+	}
+
+	if spec.Region == "" {
+		spec.Region = in.Region
+	}
+
+	spec.Network = gcp.LateInitializeString(spec.Network, in.Network)
+	spec.Description = gcp.LateInitializeString(spec.Description, in.Description)
+	spec.EnableFlowLogs = gcp.LateInitializeBool(spec.EnableFlowLogs, in.EnableFlowLogs)
+	spec.PrivateIPGoogleAccess = gcp.LateInitializeBool(spec.PrivateIPGoogleAccess, in.PrivateIpGoogleAccess)
+	if len(in.SecondaryIpRanges) != 0 && len(spec.SecondaryIPRanges) == 0 {
+		spec.SecondaryIPRanges = make([]*v1beta1.SubnetworkSecondaryRange, len(in.SecondaryIpRanges))
+		for i, r := range in.SecondaryIpRanges {
+			spec.SecondaryIPRanges[i] = &v1beta1.SubnetworkSecondaryRange{
+				IPCidrRange: r.IpCidrRange,
+				RangeName:   r.RangeName,
+			}
+		}
+	}
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(in *v1beta1.SubnetworkParameters, current compute.Subnetwork) (upToDate bool, privateAccess bool) {
```

Comment:
This works a little bit different than how we returned the update function itself for node pools or gke clusters https://github.com/crossplaneio/stack-gcp/blob/39b9ff38a6f8d355c0b1c31f99a142c91af6cbd6/pkg/clients/cluster/cluster.go#L1026

However, it seemed like a simpler solution given that there are only two options here. This way we keep the actual GCP API logic in the controller.